examples/step-37/step-37.cc

```diff
@@ -493,176 +406,319 @@ namespace Step37
+        phi.reinit (cell);
+        for (unsigned int i=0; i<phi.dofs_per_cell; ++i)
+          {
+            for (unsigned int j=0; j<phi.dofs_per_cell; ++j)
+              phi.submit_dof_value(VectorizedArray<number>(), j);
+            phi.submit_dof_value(make_vectorized_array<number>(1.), i);
+
+            phi.evaluate (false, true);
+            for (unsigned int q=0; q<phi.n_q_points; ++q)
+              phi.submit_gradient (coefficient(cell,q) *
+                                   phi.get_gradient(q), q);
+            phi.integrate (false, true);
+            diagonal[i] = phi.get_dof_value(i);
+          }
+        for (unsigned int i=0; i<phi.dofs_per_cell; ++i)
+          phi.submit_dof_value(diagonal[i], i);
+        phi.distribute_local_to_global (dst);
+      }
   }
 
 
 
-  // Regarding the calculation of the diagonal, we expect the user to provide
-  // a vector with the diagonal entries (and we will compute them in the code
-  // below). We only need it for the level matrices of multigrid, not the
-  // system matrix (since we only need these diagonals for the multigrid
-  // smoother). Since we fill only elements into unconstrained entries, we
-  // have to set constrained entries to one in order to avoid the same
-  // problems as discussed above.
-  template <int dim, int fe_degree, typename number>
-  void
-  LaplaceOperator<dim,fe_degree,number>::set_diagonal(const Vector<number> &diagonal)
+  // @sect4{MGInterfaceMatrix class}
+
+  // The adaptive multigrid realization in deal.II implements an approach
+  // called local smoothing. This means that the smoothing on the finest level
+  // only covers the local part of the mesh defined by the fixed (finest) grid
+  // level and ignores parts of the computational domain where the terminal
+  // cells are coarser than this level. As the method progresses to coarser
+  // levels, more and more of the global mesh will be covered. At some coarser
+  // level, the whole mesh will be covered. Since all level matrices in the
+  // multigrid method cover a single level in the mesh, no hanging nodes
+  // appear on the level matrices. At the interface between multigrid levels,
+  // homogeneous Dirichlet boundary conditions are set while smoothing. When
+  // the residual is transferred to the next coarser level, however, the
+  // coupling over the multigrid interface needs to be taken into account.
+  // This is done by the so-called interface (or edge) matrices that compute
+  // the part of the residual that is missed by the level matrix with
+  // homogeneous Dirichlet conditions. We refer to the @ref mg_paper
+  // "Multigrid paper by Janssen and Kanschat" for more details.
+  //
+  // For the implementation of those interface matrices, most infrastructure
+  // is already in place and provided by MatrixFreeOperators::Base through the
+  // two multiplication routines vmult_interface_down() and
+  // vmult_interface_up(). The only thing we have to do in this tutorial
```

Comment:
don't we want to add a base class for that? That's going to be used in every matrix-free application anyway.